sheet_to_triples/tests/test_xl.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2021 Visual Meaning Ltd
+# This is free software licensed as GPL-3.0-or-later - see COPYING for terms.
+
+"""Unittests for the xl, xls and xlsx modules of sheet-to-triples."""
+
+import unittest
+
+from unittest import mock
+
+from .. import xl, xlsx, xls
+
+
+class TestSheet:
+
+    def __init__(self, name):
+        self.name = name
+
+    def _get_rows(self, caller):
+        return 'test {} {}'.format(caller, self.name)
+
+    def iter_rows(self):
+        return self._get_rows('iter_rows')
+
+    def get_rows(self):
+        return self._get_rows('get_rows')
+
+
+class TestXlsBook:
+
+    def __init__(self, sheets):
+        self.sheets = sheets
+
+    def sheet_names(self):
+        return self.sheets
+
+    def sheet_by_name(self, sheet):
+        return TestSheet(sheet)
+
+
+class TestCell:
+
+    def __init__(self, value):
+        self.value = value
+
+
+class TestXL(unittest.TestCase):
+
+    _mock_xls_open = mock.patch(
+            'xlrd.open_workbook', return_value='xls_book')
+
+    _mock_xlsx_open = mock.patch(
+            'openpyxl.load_workbook', return_value='xlsx_book')
+
+    def test_load_book_xls(self):
+        with self._mock_xls_open as xls_o:
+            book = xl.load_book('testbook.xls')
+
+        self.assertIsInstance(book, xls.Book)
+        self.assertEqual(book._book, 'xls_book')
+        xls_o.assert_called_with('testbook.xls')
+
+    def test_load_book_xlsx(self):
+        with self._mock_xlsx_open as xlsx_o:
+            book = xl.load_book('testbook.xlsx')
+
+        self.assertIsInstance(book, xlsx.Book)
+        self.assertEqual(book._book, 'xlsx_book')
+        xlsx_o.assert_called_with('testbook.xlsx', data_only=True)
+
+    def test_load_book_not_excel_filetype(self):
+        with self.assertRaises(ValueError):
+            xl.load_book('testbook.bogus')
+
+    def test_iter_sheet_xlsx(self):
+        books = [
+            {'a': TestSheet('a')},
+            {'b': TestSheet('b')},
+        ]
+        xslx_books = [xlsx.Book(book) for book in books]
+        value = xl.iter_sheet(xslx_books, 'b')
+        self.assertEqual(value, 'test iter_rows b')
+
+    def test_iter_sheet_xlsx_no_matching_sheet(self):
+        books = [
+            {'a': TestSheet('a')},
+            {'b': TestSheet('b')},
+        ]
+        xlsx_books = [xlsx.Book(book) for book in books]
+        with self.assertRaises(ValueError):
+            xl.iter_sheet(xlsx_books, 'c')
+
+    def test_iter_sheet_xls(self):
+        books = [
+            TestXlsBook(['a']),
+            TestXlsBook(['b']),
+        ]
+        xls_books = [xls.Book(book) for book in books]
+        value = xl.iter_sheet(xls_books, 'b')
+        self.assertEqual(value, 'test get_rows b')
+
+    def test_iter_sheet_xls_no_matching_sheet(self):
+        books = [
+            TestXlsBook(['a']),
+            TestXlsBook(['b']),
+        ]
+        xls_books = [xls.Book(book) for book in books]
+        with self.assertRaises(ValueError):
```

Comment:
Probably worth testing the string contents of the value error here by capturing the context.

Reply:
I was initially doing this, but then went back on it because I thought that capturing `stderr` would make unrelated exceptions more difficult to debug if the test triggered them. Still, I've had to do a similar thing in the tests for `__main__` to stop it from vomiting argparse usage helper prompts into the test output, so probably worth doing it here too.

Reply:
Or I could just get the error message out of the exception object. That would be far easier.